### PR TITLE
[WIP] Support all early stopping options: `LambdaEarlyStopping`

### LINKED ISSUE #6795: Add more early stopping options
## 🚀 Feature
Additional early stopping features

@tchaton 

First, `max_time` which should probably be in parallel to `max_epochs` in the main trainer loop.  Why an additional one?  Because (1) you never have any idea how long an epoch will be - especially if you tinker with hyperparameters; and (2) sometimes you want to give an amount of time and see which version of the model does the best given a fixed amount of time.

Second, a few variations on the `EarlyStopping` callback which is based on a metric.
1. A way to stop because things have diverged completely and you doubt it can recover
2. A way to stop because things have converged completely in terms of the quality of the approximation and there is no point doing futher iterations.  This is distinct from convergence because it has ceased to get better - which is what it currently does.

For both of these, I think it is useful to have a `min_epochs` or somethign option to ensure that it doesn't stop right away.  I think that is what https://github.com/PyTorchLightning/pytorch-lightning/pull/6705  is supposed to do though so it isn't needed here?

Finally, I think that it would be great in PL to have a way to log the reason for stopping so that it can be seen in the logs and be available within grid experiments view.  Not sure the way to do that though, but maybe the callback could save a string in the logs?

## Implementation

I implemented these two features in the currrent callaback with something like:
```python
    def __init__(
        self,
        monitor: str = 'early_stop_on',
        min_delta: float = 0.0,
        patience: int = 3,
        verbose: bool = False,
        mode: str = 'min',
        strict: bool = True,
        stopping_threshold: float = 0.0,
        divergence_threshold: float = 1e6
    ):
        super().__init__()
        self.monitor = monitor
        self.patience = patience
        self.verbose = verbose
        self.strict = strict
        self.min_delta = min_delta
        self.wait_count = 0
        self.stopped_epoch = 0
        self.mode = mode
        self.warned_result_obj = False

        self.__init_monitor_mode()

        self.min_delta *= 1 if self.monitor_op == torch.gt else -1
        torch_inf = torch.tensor(np.Inf)
        self.best_score = torch_inf if self.monitor_op == torch.lt else -torch_inf

        self.stopping_threshold = stopping_threshold
        self.divergence_threshold = divergence_threshold
        self.last_time = time()
        self.elapsed_time = 0.0
```
Then I added in something like
```python
    def _run_early_stopping_check(self, trainer, pl_module):
        """
        Checks whether the early stopping condition is met
        and if so tells the trainer to stop the training.
        """
        # ADDED
        self.elapsed_time += time() - self.last_time
        self.last_time = time()

        logs = trainer.callback_metrics

        if (
            trainer.fast_dev_run  # disable early_stopping with fast_dev_run
            or not self._validate_condition_metric(logs)  # short circuit if metric not present
        ):
            return  # short circuit if metric not present

        current = logs.get(self.monitor)

        # when in dev debugging
        trainer.dev_debugger.track_early_stopping_history(self, current)

        if self.monitor_op(current - self.min_delta, self.best_score):
            self.best_score = current
            self.wait_count = 0
        else:
            self.wait_count += 1

        # ADDED
        if self.wait_count >= self.patience:
            self.stopped_epoch = trainer.current_epoch
            trainer.should_stop = True
            print(f"\n{OKCYAN}Stopping. Above patience of {self.patience} epochs without improvement of {self.min_delta}")            
        elif(self.monitor_op(current, self.stopping_threshold)):
            self.stopped_epoch = trainer.current_epoch
            trainer.should_stop = True
            print(f"\n{OKCYAN}Stopping. Below tolerance {self.monitor} = {logs[self.monitor]} <= {self.stopping_threshold}{ENDC}")          
        elif(self.monitor_op(-current,-self.divergence_threshold) or torch.isnan(current)):
            self.stopped_epoch = trainer.current_epoch
            trainer.should_stop = True
            print(f"\n{OKCYAN}Stopping. Divergence {self.monitor} = {logs[self.monitor]} >= {self.divergence_threshold} {ENDC}")      
        # stop every ddp process if any world process decides to stop
        trainer.should_stop = trainer.training_type_plugin.reduce_boolean_decision(trainer.should_stop)
```

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #6795

## Pitch

```python
inf = torch.tensor(np.Inf)  # our stop fn maximizes

def bounds_fn(score: float, _ = None) -> Tuple[bool, str]:
    return not (10 < score <= 100), "Out of bounds"

def stop_fn(score: float, best: float = inf) -> bool:
    # maximizing - did not improve
    return score < best

# stateful class example
class StdDevUnder:
    def __init__(self, threshold: float, num_values_to_keep: int) -> None:
        assert threshold > 0, "Standard deviation should be a positive value"
        assert num_values_to_keep > 1, (
            "The number of values to keep must be greater than 1 to compute "
            "the standard deviation"
        )
        self._threshold = threshold
        self._num_values_to_keep = num_values_to_keep
        self._values = []
        self._nval = 0

    def __call__(self, score: float, _ = None) -> Union[bool, Tuple[bool, str]]:
        # Add last value to the values
        if self._num_values_to_keep > len(self._values):
            self._values.append(value)
        else:
            self._values[self._nval] = value
            self._nval = (self._nval + 1) % self._num_values_to_keep

        # If not enough values are kept, return False
        if len(self._values) < self._num_values_to_keep:
            return False

        std = torch.std(torch.stack(self._values))
        if std < self._threshold:
            reason = f"Standard deviation {std} < Threshold {self._threshold}"
            return True, reason
        return False

Trainer(callbacks=[
    LambdaEarlyStopping(stop_fn),
    # stop immediately when the target cost is reached or the run failed completely
    LambdaEarlyStopping(bounds_fn, patience=0),
    # stop when the stddev over the last 10 values is lower than a threshold
    LambdaEarlyStopping(StdDevUnder(0.1, 10)),
])
```

There is no need to pass `mode=min|max`. The initial best value is provided by the user with the `best` keyword argument.

The `Callable` can optionally return a string with the stopping reason. The callable function name is also shown in the "report".

The `Callable` can be a class callable and keep a state.

## Notes

This does not replace the PR in #6868 as we still want to provide that functionality out-of-the-box.

## In a future PR

- Refactor `EarlyStopping` so it subclasses `LambdaEarlyStopping` to reuse code: The objective of EarlyStopping will be to provide out-of-the-box stopping functions.

## TODO

- Code: Opened this PR to allow people to read the pitch
- Docs
- Tests

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Comment by Borda** [comment]
seem this quite old PR with several handerts of commits behind the master, consider finishing it or closing as most likely the conflicts will make the PR challenging to finish... :rabbit:
